clap_complete/src/shells/fish.rs

```diff
@@ -57,31 +86,37 @@ fn gen_fish_inner(
     //      -a "{possible_arguments}"
     //      -r # if require parameter
     //      -f # don't use file completion
-    //      -n "__fish_use_subcommand"               # complete for command "myprog"
-    //      -n "__fish_seen_subcommand_from subcmd1" # complete for command "myprog subcmd1"
+    //      -n "{needs_fn_name}"            # complete for command "myprog"
+    //      -n "{using_fn_name} subcmd1"    # complete for command "myprog subcmd1"
 
     let mut basic_template = format!("complete -c {root_command}");
 
     if parent_commands.is_empty() {
         if cmd.has_subcommands() {
-            basic_template.push_str(" -n \"__fish_use_subcommand\"");
+            basic_template.push_str(&format!(" -n \"{needs_fn_name}\""));
         }
     } else {
-        let mut out = String::from("__fish_seen_subcommand_from");
-        for &command in parent_commands {
-            out.push(' ');
-            out.push_str(command);
-        }
-        let subcommands: Vec<&str> = cmd
-            .get_subcommands()
-            .flat_map(Command::get_name_and_visible_aliases)
-            .collect();
-        if !subcommands.is_empty() {
-            out.push_str("; and not __fish_seen_subcommand_from");
-        }
-        for name in subcommands {
-            out.push(' ');
-            out.push_str(name);
+        let mut out = String::from(using_fn_name);
+        match parent_commands {
+            [] => unreachable!(),
+            [command] => {
+                out.push_str(&format!(" {command}"));
+                if cmd.has_subcommands() {
+                    out.push_str("; and not __fish_seen_subcommand_from");
+                }
+                let subcommands = cmd
+                    .get_subcommands()
+                    .flat_map(Command::get_name_and_visible_aliases);
+                for name in subcommands {
+                    out.push_str(&format!(" {name}"));
+                }
+            }
+            [command, subcommand] => out.push_str(&format!(
+                " {command}; and __fish_seen_subcommand_from {subcommand}"
+            )),
+            // Subcommand should be nested less than 3 levels to be practical
+            // `rustup toolchain help install` cannot receive any flags to it.
```

Comment:
... this assumption should be called out in the PR description.

I also believe this deserves a `HACK` prefix

```suggestion
            // HACK: Assuming subcommands are only nested less than 3 levels as more than that is unwieldy and takes more effort to support
            // e.g. `rustup toolchain help install` cannot receive any flags to it.
```